src/udx.c

```diff
@@ -1195,46 +1195,71 @@ on_uv_poll (uv_poll_t *handle, int status, int events) {
     }
   }
 
-  while (events & UV_WRITABLE && socket->send_queue.len > 0) {
-    udx_packet_t *pkt = (udx_packet_t *) udx__fifo_shift(&(socket->send_queue));
-    if (pkt == NULL) continue;
-
-    bool adjust_ttl = pkt->ttl > 0 && socket->ttl != pkt->ttl;
+  /* start mmsg version */
+  /* alternative - send e.g. 20 at a time (see vendor/libuv/src/unix/udp.c) */
+  udx_fifo_t *sq = &(socket->send_queue);
+
+  while (events & UV_WRITABLE && sq->len > 0) {
+    unsigned int sqlen = sq->len;
+    //struct mmsghdr *h = calloc(sqlen, sizeof(struct mmsghdr));
+    udx_packet_t **batch = malloc(sqlen * sizeof(udx_packet_t *));
+    int pkts = 0;
+
+    for (int i = 0; i < sqlen; i++) {
+      udx_packet_t *pkt = udx__fifo_shift(&(socket->send_queue));
+      // asssert (*pkt != NULL && "handle null");
+      /* todo: how to trigger this condition? */
+      if (pkt == NULL) {
+        break;
+      }
```

Comment:
Given that the loop is guarded by the queue length, I don't think this should ever happen? @mafintosh might have some insights.

Reply:
I've confirmed it does occasionally happen (in master and this branch): when ack_packet calls udx__fifo_remove it leaves a null in the fifo.